src/python/pants/source/source_root.py

```diff
@@ -361,6 +366,25 @@ def _do_add_pattern(self, pattern, langs, category):
     node.category = category
     node.is_terminal = True
 
+  def traverse(self) -> Set[UncanonicalizedSourceRoot]:
+    uncanonicalized_source_roots = set()
+    lang_canonicalizations = self._source_root_factory._lang_canonicalizations
+    all_lang_names = tuple(lang_canonicalizations.keys())
+
+    def traverse_helper(node, path_components):
+      for name in node.children:
+        child = node.children[name]
+        if child.is_terminal:
+          effective_path = '/'.join([*path_components, name])
+          effective_lang_names = child.langs if len(child.langs) != 0 else all_lang_names
+          category = child.category
+          root = UncanonicalizedSourceRoot(effective_path, effective_lang_names, category)
```

Comment:
Nit. kwargs may make this even more readable:
```suggestion
          root = UncanonicalizedSourceRoot(path=effective_path, langs=effective_lang_names, category=category)
```

You could also then inline the line above to have `category=child.category` for less variables to keep track of.